### DISCUSSION TITLE
Working with XML Documents

### DISCUSSION THREAD

**Discussion by oswaldobapvicjr** [ezylang/EvalEx#481] [Q&A] [answered]
I have an explicit requirement to filter a previously loaded XML document using an XPath expression in a custom function. For example:

````
xpath(document, "/products/product[1]/price")
````

However, to do it in EvalEx, I would need the `EvaluationValue` class to accept the `DataType.DOCUMENT`, stored as `org.w3c.dom.Document` from the Java API (no additional dependency required). Could you consider having it in EvalEx?

A choice would be accepting a more generic `DataType` type (e.g.: `DataType.UNDEFINED`). The point is to take the Document as a variable inside the context so that the `xpath` function can filter it. This second choice feels more "open" to further requirements. Since `EvaluationValue.DataType` is an `enum`, it doesn't allow adding custom data types on top of it without changing the class.

Note 1: We could also accept a `org.w3c.dom.Node`, since a Document is a type of (root) Node.

Note 2: Inspired by the usability of MSExcel function: `FILTERXML(xml, xpath)`

**Comment by uklimaschewski** [comment, marked as answer]
I think that a `DataType.BINARY` would be an aceptable generic new data type for such cases.
The new converter could be last in the converter chain in `DefaultEvaluationValueConverter`, allowing to convert any object that wasn't converted before (just taking the Object and setting it as value).

**Reply by oswaldobapvicjr** [reply]
Good. I think I can implement it with a pull request.

**Reply by oswaldobapvicjr** [reply]
I just created the Pull Request #482 with this implementation. ;)
